crates/bevy_utils/src/intern.rs

```diff
@@ -0,0 +1,291 @@
+//! Provides types used to statically intern immutable values.
+//!
+//! Interning is a pattern used to save memory by deduplicating identical values,
+//! speed up code by shrinking the stack size of large types,
+//! and make comparisons for any type as fast as integers.
+
+use std::{
+    fmt::Debug,
+    hash::Hash,
+    ops::Deref,
+    ptr,
+    sync::{OnceLock, PoisonError, RwLock},
+};
+
+use crate::HashSet;
+
+/// An interned value. Will stay valid until the end of the program and will not drop.
+///
+/// For details on interning, see [the module level docs](self).
+///
+/// # Implementation details
+///
+/// This is just a reference to a value with an `'static` lifetime.
+//
+// This implements [`Copy`], [`Clone`], [`PartialEq`], [`Eq`] and [`Hash`] regardles of what `T`
+// implements, because it only uses the pointer to the value and not the value itself.
+// Therefore it MUST NEVER implement [`Borrow`](`std::borrow::Borrow`).
```

Comment:
```suggestion
/// # Comparisons
/// 
/// Interned values use reference equality, meaning they implement [`Eq`]
/// and [`Hash`] regardless of whether `T` implements these traits.
/// Two interned values will only compare equal if they were interned using
/// the same [`Interner`] instance.
// NOTE: This type must NEVER implement Borrow since it does not obey that trait's invariants.
///
/// ```
/// # use bevy_utils::intern::*;
/// #[derive(PartialEq, Eq, Hash)]
/// struct Value(i32);
/// impl Leak for Value {
///     // ...
/// # fn leak(&self) -> &'static Self { Box::leak(Box::new(Value(self.0))) }
/// }
/// let interner_1 = Interner::new();
/// let interner_2 = Interner::new();
/// // Even though both values are identical, their interned forms do not
/// // compare equal as they use different interner instances.
/// assert_ne!(interner_1.intern(&Value(42)), interner_2.intern(&Value(42)));
/// ```
```

I think this behavior is important to document and test for. Feel free to tweak the phrasing.